### PR TITLE
Implement Group Stage end

### PR DESCRIPTION
**Pull request by Malaber**

### CODE REVIEW

File: app/services/group_stage_service.rb [hunk 1]
```diff
@@ -55,5 +55,35 @@ def update_group_scores(group)
...
+
+    # Returns a list of the teams in the group sorted by their group_points, difference_in_points, scored_points
+    #
+    # @param group Group the group to get the teams from
+    # @return [Array] of teams
+    def teams_sorted_by_group_scores(group)
+      group.teams.sort do |a, b|
+        [group.group_scores.find_by(team: b).group_points,
+         group.group_scores.find_by(team: b).difference_in_points,
+         group.group_scores.find_by(team: b).scored_points] <=>
```
Comment (Thor77) [inline app/services/group_stage_service.rb]: please assign the return value of `...find_by()` to a var, they cause a db query every time...

File: spec/controllers/stages_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      expect(Stage.find_by(id: body[:id])).to eq(@stage)
+      expect(body[:level]).to eq(@stage.level)
+      expect(body[:state]).to eq(@stage.state)
+    end
+  end
+
+  describe 'PUT #update' do
+    context 'group_stage with matches that are done' do
+      before do
+        @running_group_stage = create(:group_stage, match_factory: :finished_group_match)
```
Comment (Thor77) [inline spec/controllers/stages_controller_spec.rb]: please use let instead of class vars

File: spec/controllers/stages_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      end
+
+      FINISHED = { state: 'finished' }.freeze
```
Comment (Thor77) [inline spec/controllers/stages_controller_spec.rb]: why is this a constant instead of a let statement?

File: spec/factories/groups.rb [hunk 4]
```diff
@@ -7,7 +7,7 @@
       match_factory { :group_match }
     end
 
-    sequence(:number)
+    number { rand(10) }
```
Comment (Thor77) [inline spec/factories/groups.rb]: will it cause problems if there are multiple groups with the same number?
what's the problem with a sequence here?

**Review by Thor77 — CHANGES_REQUESTED** [review]
Generally I think the controller includes far too much business logic here that should live in a service. Controller tests should then mock this service to not test the same thing over and over again. But as there's not much time left.. we can look at this somewhen in the future.. or never

File: spec/controllers/stages_controller_spec.rb (hunk 2 above)
Comment (Malaber) [reply spec/controllers/stages_controller_spec.rb]: done

File: spec/controllers/stages_controller_spec.rb (hunk 3 above)
Comment (Malaber) [reply spec/controllers/stages_controller_spec.rb]: because random.org

File: spec/factories/groups.rb (hunk 4 above)
Comment (Malaber) [reply spec/factories/groups.rb]: i wanted to change things :(

File: app/services/group_stage_service.rb (hunk 1 above)
Comment (Malaber) [reply app/services/group_stage_service.rb]: done

**Review by Thor77 — APPROVED** [review]

**Pull request merged**